executor/window.go

```diff
@@ -0,0 +1,174 @@
+	"github.com/pingcap/tidb/util/chunk"
+)
+
+// WindowExec is the executor for window functions. Note that it only supports aggregation without frame clause now.
+type WindowExec struct {
+	baseExecutor
+
+	groupChecker  *groupChecker
+	inputIter     *chunk.Iterator4Chunk
+	inputRow      chunk.Row
+	groupRows     []chunk.Row
+	childResults  []*chunk.Chunk
+	windowFunc    aggfuncs.AggFunc
+	partialResult aggfuncs.PartialResult
+	resultColIdx  int
+	executed      bool
+	meetNewGroup  bool
+	remainingRows int64
+}
+
+// Close implements the Executor Close interface.
+func (e *WindowExec) Close() error {
+	e.childResults = nil
+	return errors.Trace(e.baseExecutor.Close())
+}
+
+// Next implements the Executor Next interface.
+func (e *WindowExec) Next(ctx context.Context, chk *chunk.Chunk) error {
+	if span := opentracing.SpanFromContext(ctx); span != nil && span.Tracer() != nil {
+		span1 := span.Tracer().StartSpan("windowExec.Next", opentracing.ChildOf(span.Context()))
+		defer span1.Finish()
+	}
+	if e.runtimeStats != nil {
+		start := time.Now()
+		defer func() { e.runtimeStats.Record(time.Now().Sub(start), chk.NumRows()) }()
+	}
+	chk.Reset()
+	if e.meetNewGroup && e.remainingRows > 0 {
+		err := e.appendResult2Chunk(chk)
+		if err != nil {
+			return err
+		}
+	}
+	for !e.executed && (chk.NumRows() == 0 || chk.RemainedRows(e.resultColIdx) > 0) {
+		err := e.consumeOneGroup(ctx, chk)
+		if err != nil {
+			e.executed = true
+			return errors.Trace(err)
+		}
+	}
+	return nil
+}
+
+func (e *WindowExec) consumeOneGroup(ctx context.Context, chk *chunk.Chunk) error {
+	var err error
+	if err = e.fetchChildIfNecessary(ctx, chk); err != nil {
+		return errors.Trace(err)
+	}
+	for ; e.inputRow != e.inputIter.End(); e.inputRow = e.inputIter.Next() {
+		e.meetNewGroup, err = e.groupChecker.meetNewGroup(e.inputRow)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if e.meetNewGroup {
+			err := e.consumeGroupRows(chk)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			err = e.appendResult2Chunk(chk)
+			if err != nil {
+				return errors.Trace(err)
+			}
+		}
+		e.groupRows = append(e.groupRows, e.inputRow)
+		if e.meetNewGroup {
+			e.inputRow = e.inputIter.Next()
+			return nil
+		}
+	}
+	return nil
+}
+
+func (e *WindowExec) consumeGroupRows(chk *chunk.Chunk) error {
+	if len(e.groupRows) == 0 {
+		return nil
+	}
+	err := e.windowFunc.UpdatePartialResult(e.ctx, e.groupRows, e.partialResult)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	e.remainingRows += int64(len(e.groupRows))
+	e.groupRows = e.groupRows[:0]
+	return nil
+}
+
+func (e *WindowExec) fetchChildIfNecessary(ctx context.Context, chk *chunk.Chunk) (err error) {
+	if e.inputIter != nil && e.inputRow != e.inputIter.End() {
+		return nil
+	}
+
+	// Before fetching a new batch of input, we should consume the last groupChecker.
+	err = e.consumeGroupRows(chk)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	childResult := e.children[0].newFirstChunk()
```

Comment:
`tidb_max_chunk_size` may result in unnecessary results.

Reply:
Yeah, to avoid unnecessary results, we should use a smaller chunk size for the first `Chunk` used to fetch child result. But in the next time, we should use a bigger chunk size. This function always use `newFirstChunk()` to fetch the child result, it's not efficient for big tables.

Reply:
I tried the way mentioned in the comment, currently, we need to call `chunk.Renew` to create a larger chunk, and it requires to remember the last chunk... Maybe we can refine it after we can pass down the chunk size to the child.

Reply:
Okay.

Reply:
Could you file an issue to trace this problem?

Reply:
Done, https://github.com/pingcap/tidb/issues/9045